lib/Conversion/CalyxToFSM/CalyxToFSM.cpp

```diff
@@ -268,6 +268,71 @@ LogicalResult CompileFSMVisitor::visit(StateOp currentState, EnableOp enableOp,
   return success();
 }
 
+// CompileInvoke is used to convert invoke operations to group operations and
+// enable operations.
+class CompileInvoke {
+public:
+  CompileInvoke(ComponentOp component, OpBuilder builder)
+      : component(component), builder(builder) {}
+  void compile();
+
+private:
+  void lowerInvokeOp(InvokeOp invokeOp);
+  ComponentOp component;
+  OpBuilder builder;
+  // The groupNameMap is used to help generate the names of calyx.group,
+  // for example, if key is reg, value is 0, then calyx.group is named
+  // calyx.group invoke_reg_0. When an invoke is encountered, if the
+  // callee is not in the groupNameMap, then the callee is added to
+  // the groupNameMap and the corresponding value is made to be 0.
+  llvm::StringMap<size_t> groupNameMap;
+};
+
+// Access all invokeOp.
+void CompileInvoke::compile() {
+  llvm::SmallVector<InvokeOp> invokeOps =
+      component.getControlOp().getInvokeOps();
+  for (InvokeOp op : invokeOps) {
+    lowerInvokeOp(op);
+  }
+}
+
+// Convert an invoke operation to a group operation and an enable operation.
+void CompileInvoke::lowerInvokeOp(InvokeOp invokeOp) {
+  // Create a ConstantOp to assign a value to the go port.
+  Operation *prevNode = component.getWiresOp().getOperation()->getPrevNode();
+  builder.setInsertionPointAfter(prevNode);
+  hw::ConstantOp constantOp = builder.create<hw::ConstantOp>(
+      prevNode->getLoc(), builder.getI1Type(), 1);
+  Location loc = component.getWiresOp().getLoc();
+  // Set the insertion point at the end of the wires block.
+  builder.setInsertionPointToEnd(component.getWiresOp().getBodyBlock());
+  llvm::StringRef callee = invokeOp.getCallee();
+  std::string groupNameHead = "invoke_" + callee.str() + "_";
+  size_t &groupNameTail = groupNameMap[callee];
+  std::string groupName = groupNameHead + std::to_string(groupNameTail);
+  // Check if the group name is correct.
+  while (component.getWiresOp().lookupSymbol(groupName))
+    groupName = groupNameHead + std::to_string(++groupNameTail);
+  ++groupNameTail;
+  GroupOp groupOp = builder.create<GroupOp>(loc, groupName);
+  builder.setInsertionPointToStart(groupOp.getBodyBlock());
+  Value go = invokeOp.getInstGoValue();
+  // Assign a value to the go port.
+  builder.create<AssignOp>(loc, go, constantOp);
+  auto ports = invokeOp.getPorts();
+  auto inputs = invokeOp.getInputs();
+  // Generate a series of assignment operations from a list of parameters.
+  for (auto [port, input] : llvm::zip(ports, inputs))
+    builder.create<AssignOp>(loc, port, input);
+  Value done = invokeOp.getInstDoneValue();
+  // Generate a group_done operation with the instance's done port.
```

Comment:
One minor nitpick here: the comments are great, but rather than a wall of text, I think it would be significantly more readable if you put a newline before each comment line.